### PR TITLE
[SPARK-2788] [STREAMING] Add location filtering to Twitter streams

### PR DESCRIPTION
**Pull request by sjbrunst**
TwitterUtils.createStream(...) allows users to specify keywords that restrict the tweets that are returned. This change adds a location parameter that also restricts the returned tweets.

closes #2098

### CODE REVIEW

**Comment by philcontrolf1** [comment]
Hi all - this functionality is certainly something I'm interested in, but this discussion seems to have stalled somewhat. Is this still the "main" discussion for geofiltered tweets, or has it moved somewhere else?

If it's the main discussion, what needs to happen to get this moving again? I'm more than happy to write code if necessary :-)

**Comment by sjbrunst** [comment]
The discussion hasn't moved anywhere, as far as I know. I was waiting for @tdas to look at the latest changes.

**Comment by JoshRosen** [comment]
To throw a monkey wrench into this discussion: do we really want to be maintaining a Twitter library inside of Spark itself or should we try to move the ongoing development of this source into a separate third-party package?

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterInputDStream.scala [hunk 1]
```diff
@@ -28,11 +28,16 @@ import org.apache.spark.storage.StorageLevel
 import org.apache.spark.Logging
 import org.apache.spark.streaming.receiver.Receiver
 
-/* A stream of Twitter statuses, potentially filtered by one or more keywords.
+/** Bounding box to filter tweets by location. Units are degrees. */
+case class BoundingBox(west: Double, south: Double, east: Double, north: Double)
```
Comment (huitseeker) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterInputDStream.scala]: What if I want to filter using a non-contiguous surface (e.g. I want to know who talks about soccer in Manchester or Chelsea during a football match betw. their 2 teams) ?

File: external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterInputDStream.scala [hunk 2]
```diff
@@ -86,12 +97,18 @@ class TwitterReceiver(
...
+      query.count(count)
+      if (follow.size > 0) {
+        query.follow(follow.toArray)
+      }
+      if (track.size > 0) {
+        query.track(track.toArray)
+      }
+      if (locations.size > 0) {
+        query.locations(locations.flatMap(box =>
+          Seq(Array(box.west, box.south), Array(box.east, box.north))).toArray)
```
Comment (huitseeker) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterInputDStream.scala]: Why not `Array(Array(box.west, box.south), Array(box.east, box.north))` ?

Comment (sjbrunst) [inline external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterInputDStream.scala]: Good catch. It won't compile without the `.toArray` at the end, but I can change the `Seq` to `Array`. I will update the code.

**Comment by dmvieira** [comment]
Hey guys, I need this patch too.

**Comment by srowen** [comment]
I don't think this PR is going forward, and should be closed. Do you mind closing this PR @sjbrunst ? You can see some other related efforts, but the impression I have from many related discussions is that this belongs outside Spark.

**Comment by dmvieira** [comment]
But without this path you're restricting a lot Twitter functionalities inside Spark and still supporting Twitter interface. Spark still maintain Twitter API interface even without this path. IMHO if Spark don't want to maintain Twitter interface you should remove Twitter streaming as a package inside Spark

**Comment by sjbrunst** [comment]
@srowen I am willing to close this PR, but I agree with @dmvieira . It doesn't matter to me whether this functionality gets added inside Spark or as an external library, but either way it should go somewhere because there is enough demand for it. This patch just expands on the Twitter library that is already part of Spark. If there are plans to make the Twitter library external to Spark, then this change (or a similar PR) can move along with it.

**Comment by JoshRosen** [comment]
Personally I would love to see the Twitter package moved from Spark itself into a separate project / package; the only reason that we have it is for legacy reasons.

**Comment by dmvieira** [comment]
So, why not improve it with this PR and then move it to a new project / package when we think about a better solution? We can create an issue or you can talk with stakeholders to discuss about it.

**Comment by srowen** [comment]
@dmvieira I'm not sure who you're addressing there, but if this isn't something that will go in Spark, there's no reason to discuss the change here, right? Make a new github repo. Maybe that's what you mean.

**Comment by sjbrunst** [comment]
What is the timeline on moving the Twitter package out of Spark and into a separate project? If it is going to be another year before that actually happens then it might be worth it to finish this PR so users have a way to use this feature until then.

**Comment by JoshRosen** [comment]
Should we deprecate these Twitter APIs in order to encourage them to be split into a third-party package?

**Comment by dmvieira** [comment]
I'm starting a third-party package as suggested by @srowen and I hope you enjoy. Feel free to collaborate: https://github.com/dmvieira/spark-twitter-stream-receiver

**Comment by sjbrunst** [comment]
@JoshRosen Do we really want to deprecate the Twitter APIs before there is a user-friendly way to use it in an external package? I think there is a large user base for this feature, and it is a motivating example for the use of Spark Streaming in the programming guide and the code examples. I've also seen AMP Camp exercises built around this Twitter package. If this package is so heavily featured in teaching Spark Streaming and demonstrating its applications, then I think it would be strange to deprecate it before there is an easy-to-use alternative.

**Comment by srowen** [comment]
You can add any third-party package with `--packages` or add it to your application as a dependency. I don't think Josh is saying it _will_ be deprecated, certainly not now. But this PR should be closed in favor of continuing development elsewhere. That's a separate issue.

**Comment by agsachin** [comment]
@srowen Is anyone working on third-party package for twitter,

I suggest to add FilterQuery support to existing api, and move twitter Api's out of spark whenever a bigger changes is done on twitter4j api side till that time we can have twitter in spark, continuing existing api with some enhancements.

**Pull request closed without merging**